src/features/duckplayer/thumbnails.js

```diff
@@ -0,0 +1,310 @@
+ */
+
+/**
+ * This features covers the implementation
+ */
+export class Thumbnails {
+    sideEffects = new SideEffects()
+    /**
+     * @param {ThumbnailParams} params
+     */
+    constructor (params) {
+        this.settings = params.settings
+        this.messages = params.messages
+        this.environment = params.environment
+    }
+
+    /**
+     * Perform side effects
+     */
+    init () {
+        this.sideEffects.add('showing overlays on hover', () => {
+            const { selectors } = this.settings
+            const parentNode = document.documentElement || document.body
+
+            // create the icon & append it to the page
+            const icon = new IconOverlay()
+            icon.appendHoverOverlay((href) => {
+                this.messages.openDuckPlayer(new OpenInDuckPlayerMsg({ href }))
+            })
+
+            // remember when a none-dax click occurs - so that we can avoid re-adding the
+            // icon whilst the page is navigating
+            let clicked = false
+
+            // detect all click, if it's anywhere on the page
+            // but in the icon overlay itself, then just hide the overlay
+            const clickHandler = (e) => {
+                const overlay = icon.getHoverOverlay()
+                if (overlay?.contains(e.target)) {
+                    // do nothing here, the click will have been handled by the overlay
+                } else if (overlay) {
+                    clicked = true
+                    icon.hideOverlay(overlay)
+                    icon.hoverOverlayVisible = false
+                    setTimeout(() => {
+                        clicked = false
+                    }, 0)
+                }
+            }
+
+            parentNode.addEventListener('click', clickHandler, true)
+
+            const removeOverlay = () => {
+                const overlay = icon.getHoverOverlay()
+                if (overlay) {
+                    icon.hideOverlay(overlay)
+                    icon.hoverOverlayVisible = false
+                }
+            }
+
+            const appendOverlay = (element) => {
+                if (element && element.isConnected) {
+                    icon.moveHoverOverlayToVideoElement(element)
+                }
+            }
+
+            // detect hovers and decide to show hover icon, or not
+            const mouseOverHandler = (e) => {
+                if (clicked) return
+                const hoverElement = findElementFromEvent(selectors.thumbLink, selectors.hoverExcluded, e)
+                const validLink = isValidLink(hoverElement, selectors.excludedRegions)
+
+                // if it's not an element we care about, bail early and remove the overlay
+                if (!hoverElement || !validLink) {
+                    return removeOverlay()
+                }
+
+                // ensure it doesn't contain sub-links
+                if (hoverElement.querySelector('a[href]')) {
```

Comment:
Can this be configured remotely?

- In fact can't isValidLink already handle this? Does it belong there if we need it.
- Same for the img check too.

Reply:
no, because isValidLink is also used in situations where you don't have an img (regular links).

These 2 checks are explicitly here because they only need to occur when we're trying to decide if dax should be shown

Reply:
I still think we should make these more configurable, allowlist+denylist selector or something.

Reply:
As we spoke about in MM - going to follow with making more things configurable once this lands. We've added tons of ways to mitigate breakage :)